### PR TITLE
ScanAndSort query fails with NPE for simple queries

### PR DESCRIPTION
**Pull request by kgyrtkirk**
### Issue

for queries like:
```
select * from (select * from "wikipedia" limit 1000) order by "user"
```

<details>
<summary>an Exception like this is thrown</summary>

```
java.lang.NullPointerException
	at org.apache.druid.query.operator.WindowOperatorQueryQueryRunnerFactory$1$1.apply(WindowOperatorQueryQueryRunnerFactory.java:112)
	at org.apache.druid.query.operator.WindowOperatorQueryQueryRunnerFactory$1$1.apply(WindowOperatorQueryQueryRunnerFactory.java:1)
	at org.apache.druid.java.util.common.guava.MappingYieldingAccumulator.accumulate(MappingYieldingAccumulator.java:61)
	at org.apache.druid.query.operator.OperatorSequence$MyReceiver.push(OperatorSequence.java:146)
	at org.apache.druid.query.operator.ScanOperator$1.push(ScanOperator.java:105)
	at org.apache.druid.query.operator.SegmentToRowsAndColumnsOperator.goOrContinue(SegmentToRowsAndColumnsOperator.java:62)
	at org.apache.druid.query.operator.ScanOperator.goOrContinue(ScanOperator.java:77)
	at org.apache.druid.query.operator.OperatorSequence.toYielder(OperatorSequence.java:76)
	at org.apache.druid.java.util.common.guava.MappedSequence.toYielder(MappedSequence.java:49)
	at org.apache.druid.java.util.common.guava.ConcatSequence.makeYielder(ConcatSequence.java:85)
	at org.apache.druid.java.util.common.guava.ConcatSequence.toYielder(ConcatSequence.java:65)
	at org.apache.druid.java.util.common.guava.LazySequence.toYielder(LazySequence.java:46)
	at org.apache.druid.java.util.common.guava.WrappingSequence$2.get(WrappingSequence.java:88)
	at org.apache.druid.java.util.common.guava.WrappingSequence$2.get(WrappingSequence.java:1)
	at org.apache.druid.java.util.common.guava.SequenceWrapper.wrap(SequenceWrapper.java:55)
	at org.apache.druid.java.util.common.guava.WrappingSequence.toYielder(WrappingSequence.java:83)
	at org.apache.druid.java.util.common.guava.Yielders.each(Yielders.java:32)
	at org.apache.druid.query.operator.SequenceOperator.goOrContinue(SequenceOperator.java:52)
	at org.apache.druid.query.operator.NaiveSortOperator.goOrContinue(NaiveSortOperator.java:51)
	at org.apache.druid.query.operator.Operator.go(Operator.java:59)
	at org.apache.druid.query.operator.OperatorSequence.accumulate(OperatorSequence.java:64)
	at org.apache.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43)
	at org.apache.druid.java.util.common.guava.ConcatSequence.accumulate(ConcatSequence.java:42)
	at org.apache.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43)
	at org.apache.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50)
	at org.apache.druid.query.CPUTimeMetricQueryRunner$1.wrap(CPUTimeMetricQueryRunner.java:77)
	at org.apache.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45)
	at org.apache.druid.java.util.common.guava.WrappingSequence$1.get(WrappingSequence.java:50)
	at org.apache.druid.java.util.common.guava.SequenceWrapper.wrap(SequenceWrapper.java:55)
	at org.apache.druid.java.util.common.guava.WrappingSequence.accumulate(WrappingSequence.java:45)
	at org.apache.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43)
	at org.apache.druid.java.util.common.guava.Sequence.toList(Sequence.java:87)
	at org.apache.druid.sql.calcite.QueryTestRunner$ExecuteQuery.runQuery(QueryTestRunner.java:287)
	at org.apache.druid.sql.calcite.QueryTestRunner$ExecuteQuery.run(QueryTestRunner.java:264)
	at org.apache.druid.sql.calcite.QueryTestRunner.run(QueryTestRunner.java:707)
	at org.apache.druid.sql.calcite.QueryTestBuilder.run(QueryTestBuilder.java:318)
	at org.apache.druid.sql.calcite.CalciteQueryTest.testScanAndSortNPE(CalciteQueryTest.java:15125)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:258)
	at org.apache.druid.sql.calcite.util.QueryLogHook$1.evaluate(QueryLogHook.java:113)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:98)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
```

</details>

### Reason

going backwards; it comes from:

* output of `LazilyDecoratedRowsAndColumns` doesn't contain the column its not being materialized as the column is not found
* `rowSignature` contains all columns; but `null` types around the creation of  the `StorageAdapter`  in `ArrayListRowsAndColumns
* `ArrayListSegment` has the same kind of rowSignature 
* `inlineDataSource.getRowSignature()` returns this kind of signature in `InlineSegmentWrangler`
* `ClientQuerySegmentWalker` passes the signature along
* `ScanQueryQueryToolChest#resultArraySignature` comes up with the signature
* `ScanQueryQueryToolChest#getDataSourceColumnType` can’t handle `TableDataSource` so it returns `null` for all columns…

### Options 

* enable `TableDatasource`  to supply a `rowSignature`
  * in general probably `Datasource` should have that
* enhance `ScanQuery` to retain a valid `RowSignature` somehow

### Choosen approach

I've choosen the second approach for now; as the first might be a broader change
* the `ScanQuery` already had the name of the selected `columns`
* added `columnTypes` list to be abke to promote the list of column to `RowSignature`
* if both of them are available - a `RowSignature` is constructed and the above issue goes away

#### Release note
Fix NPE raised when only an order was applied to a result of subquery with limit.

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java [hunk 1]
```diff
@@ -1687,10 +1688,27 @@ private Map<String, Object> withScanSignatureIfNeeded(
     if (!plannerContext.featureAvailable(EngineFeature.SCAN_NEEDS_SIGNATURE)) {
```
Comment (gianm) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java:1688]: Please also make a change in this patch so MSQ uses `scanQuery.getRowSignature()` if `CTX_SCAN_SIGNATURE` is not available. That provides us a path to get rid of `EngineFeature.SCAN_NEEDS_SIGNATURE` in the future.

Please also add a comment here about how it's kind of a legacy thing now, replaced by `columnTypes` in the scan query object.

File: processing/src/main/java/org/apache/druid/query/scan/ScanQuery.java [hunk 2]
```diff
@@ -674,4 +688,20 @@ public boolean equals(Object obj)
     }
   }
 
+  /**
+   * Returns the RowSignature (if available)
+   */
+  @Nullable
+  public RowSignature getRowSignature()
+  {
+    if (columns != null && columnTypes != null && columns.size() == columnTypes.size()) {
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/query/scan/ScanQuery.java:697]: `columns.size() != columnTypes.size()` should be an error in the constructor

**Review by gianm — COMMENTED** [review]
Concept looks good to me, thanks!

File: processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java [hunk 3]
```diff
@@ -178,6 +178,10 @@ public QueryRunner<ScanResultValue> preMergeQueryDecoration(final QueryRunner<Sc
   @Override
   public RowSignature resultArraySignature(final ScanQuery query)
   {
+    RowSignature rowSignature = query.getRowSignature();
```
[the remark this answers is not in the document]
Comment (kgyrtkirk) [reply processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java:181]: didn't wanted to make that yet :D but it makes sense
I've moved most of it but to fully remove this method I think the best would be to remove legacy support (opened:#15918)

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 4]
```diff
@@ -15104,6 +15104,27 @@ public void testInGroupByOrderByLimitOutGroupByOrderByLimit()
         .run();
   }
 
+
+  @Test
+  public void testScanAndSortCanGetSchemaFromScanQuery()
+  {
+    skipVectorize();
+    cannotVectorize();
+    msqIncompatible();
```
Comment (gianm) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:15113]: Why is it `msqIncompatible()`? Seems like this query should work in MSQ.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 5]
```diff
@@ -15104,6 +15104,27 @@ public void testInGroupByOrderByLimitOutGroupByOrderByLimit()
         .run();
   }
 
+
+  @Test
+  public void testScanAndSortCanGetSchemaFromScanQuery()
+  {
+    skipVectorize();
+    cannotVectorize();
```
Comment (gianm) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:15112]: Shouldn't have both of these: "skip" means skip the vectorization test completely, "cannot" means run the test and expect it to fail. Generally "skip" is best if the test fundamentally doesn't make sense under vectorization, and "cannot" is best if it makes sense in principle, but just can't vectorize today due to missing some implementation. That way, when someone adds the implementation in the future, the test here will automatically fail, and that future person will know to remove the `cannotVectorize()` line.

Perhaps we should change the test framework to make it an error to set both?

**Review by gianm — COMMENTED** [review]
Main code looks good to me, just had comments on the test.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java (hunk 4 above)
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java]: I've just copied an existing testcase when I created the test - removed these flags

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java (hunk 5 above)
Comment (kgyrtkirk) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java]: yes - I'll change the testframework when I get to it ...  actually `cannotVectorize()` has a wierd issue: #15423

**Comment by kgyrtkirk** [comment]
@gianm I had to relax some checks:
I've added a configuration knob to set `rowSignatureMode` to `STRICT/ALLOW_NULL/RESORT_TO_STRING` modes.
This last one seemed like something which could be usefull as it will be more permissive - but I've made the `ALLOW_NULL` the default.
The original code behaved like `STRICT` ; however we do see `NULL` types in some cases: especially when the actual type supposed to be `NULL` -  its not possible to differentiate between valid but type is null and invalid column.

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java [hunk 6]
```diff
@@ -166,6 +169,56 @@ public String toString()
     }
   }
 
+  @SuppressWarnings("unused") // Used by Jackson serialization
+  public enum RowSignatureMode
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:173]: What do we need this for? If it's always going to be set to `ALLOW_NULL` in typical operation (as seems the case) then it's better not to have the extra code, extra context parameter, etc.

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java [hunk 7]
```diff
@@ -166,6 +169,56 @@ public String toString()
...
+    RESORT_TO_STRING {
+      @Override
+      public ColumnType getTypeForNullColumn(String columnName)
+      {
+        return ColumnType.STRING;
+      }
+    };
+
+    @JsonCreator
+    public static Vectorize fromString(String str)
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:207]: Looks like copy-paste error.

File: processing/src/main/java/org/apache/druid/segment/column/RowSignature.java [hunk 8]
```diff
@@ -392,4 +392,34 @@ public enum Finalization
      */
     UNKNOWN
   }
+
+  /**
+   * Builds a safe {@link RowSignature}.
+   *
+   * The new rowsignature will not contain `null` types - they will be replaced by STRING.
+   */
+  public RowSignature buildSafeSignature(ImmutableList<String> requestedColumnNames)
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/segment/column/RowSignature.java:401]: Doesn't seem like this really needs to be an `ImmutableList`. A `List` would do.

This method also seems very specific to what `DruidQuery` in the SQL planner needs, and IMO makes more sense as a `static` there.

**Review by gianm — APPROVED** [review]

**Pull request merged**